kliff/trainer/torch_trainer.py

```diff
@@ -0,0 +1,684 @@
+import os
+import tarfile
+from copy import deepcopy
+from typing import TYPE_CHECKING, Union
+
+import libdescriptor as lds
+import numpy as np
+import torch
+from loguru import logger
+from monty.dev import requires
+from torch.utils.data import DataLoader as TorchDataLoader
+from torch_scatter import scatter_add
+
+from .base_trainer import Trainer, TrainerError
+from .torch_trainer_utils.dataloaders import DescriptorDataset
+
+if TYPE_CHECKING:
+    from kliff.transforms.configuration_transforms import Descriptor
+
+
+class DNNTrainer(Trainer):
+    """
+    This module is used to train a descriptor based dense neural network (DNN), and deploy it
+    using TorchML framework. For using the DUNN model driver please use the legacy Loss optimizer
+    for now. In the future, the ~kliff.trainer.torch_trainer.DNNTrainer will also support both
+    loading and exporting the NN models for DUNN model driver.
+
+    Args:
+        configuration (dict): A dictionary containing the configuration for the trainer.
+        model (torch.nn.Module): A torch model to be trained. If not provided, the model
+            will be loaded from the model manifest. For model manifest based loading
+            the model must be a torchscript model, or valid TorchML model (tar or dir).
+    """
+
+    def __init__(self, configuration: dict, model=None):
+        self.configuration_transform: "Descriptor" = (
+            None  # for type checking the functions
+        )
+        self.torchscript_file = None
+        self.train_dataloader = None
+        self.validation_dataloader = None
```

Comment:
Do we want to add test_dataloader and test the model at the end of training?

Reply:
Yes, I will add it uniformly across all trainers after finalizing this release. Idea is to have a base method `test()` that can run some tests on the test datasets. It will do simple energy and forces test but will not be limited to it, but could also leverage openkim tests if user requests.